src/toxcore/dht_new/kbucket.rs

```diff
@@ -0,0 +1,857 @@
+/*
+    Copyright (C) 2013 Tox project All Rights Reserved.
+    Copyright © 2016-2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+Structure for holding nodes.
+
+Number of nodes it can contain is set during creation. If not set (aka `None`
+is supplied), number of nodes defaults to [`BUCKET_DEFAULT_SIZE`]
+(./constant.BUCKET_DEFAULT_SIZE.html).
+
+Nodes stored in `Bucket` are in [`PackedNode`](./struct.PackedNode.html)
+format.
+
+Used in [`Kbucket`](./struct.Kbucket.html) for storing nodes close to given
+PK; and additionally used to store nodes closest to friends.
+
+[Spec definition](https://zetok.github.io/tox-spec#updating-k-buckets).
+*/
+
+// ↓ FIXME expand doc
+/*! DHT part of the toxcore.
+    * takes care of the K-Bucket 
+*/
+use toxcore::crypto_core::*;
+use toxcore::dht_new::packet::PackedNode;
+use std::cmp::{Ord, Ordering};
+
+/** Calculate the [`k-bucket`](./struct.Kbucket.html) index of a PK compared
+to "own" PK.
+
+According to the [spec](https://zetok.github.io/tox-spec#bucket-index).
+
+Fails (returns `None`) if supplied keys are the same.
+*/
+pub fn kbucket_index(&PublicKey(ref own_pk): &PublicKey,
+                     &PublicKey(ref other_pk): &PublicKey) -> Option<u8> {
+
+    debug!(target: "KBucketIndex", "Calculating KBucketIndex for PKs.");
+    trace!(target: "KBucketIndex", "With PK1: {:?}; PK2: {:?}", own_pk, other_pk);
+
+    let xoring = own_pk.iter().zip(other_pk.iter()).map(|(x, y)| x ^ y); // work from shur
+    for (i, byte) in xoring.enumerate() {
+        for j in 0..8 {
+            if byte & (0x80 >> j) != 0 {
+                return Some(i as u8 * 8 + j);
+            }
+        }
+    }
+    None  // PKs are equal
+}
+
+/// Trait for functionality get `PublicKey` from PackedNode.
+pub trait GetPk {
```

Comment:
There is no reason to add this impl. Better implement method `fn pk(...) -> ...` inside `impl for PackedNode`

Reply:
I agree.
But, there is no impl for PackedNode now.
packet.rs is not proper place.
Is kbucket.rs proper place for impl PackedNode?

Reply:
You may create separated file `packed_node.rs` with `PackedNode`, `impl PackedNode { fn pk() ... }` and serializer + deserializer. It is better than `use toxcore::dht_new::packet::PackedNode;`